exonum/src/runtime/rust/mod.rs

```diff
@@ -0,0 +1,174 @@
+use crate::storage::Fork;
+
+use self::service::SystemService;
+
+#[derive(Debug, Default)]
+struct RustRuntime {
+    inner: RefCell<RustRuntimeInner>,
+}
+
+impl RustRuntime {
+    fn add_service(&self, artifact: RustArtifactSpec, service: Box<dyn SystemService>) {
+        self.inner.borrow_mut().services.insert(artifact, service);
+    }
+}
+
+#[derive(Debug, Default)]
+struct RustRuntimeInner {
+    services: HashMap<RustArtifactSpec, Box<dyn SystemService>>,
+    deployed: HashSet<RustArtifactSpec>,
+    initialized: HashMap<ServiceInstanceId, Box<dyn SystemService>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RustArtifactSpec {
+    pub name: String,
+    pub version: (u32, u32, u32),
+}
+
+impl RuntimeEnvironment for RustRuntime {
+    fn start_deploy(&self, artifact: ArtifactSpec) -> Result<(), DeployError> {
+        let artifact = if let ArtifactSpec::Rust(artifact) = artifact {
+            artifact
+        } else {
+            return Err(DeployError::WrongArtifact);
+        };
+
+        let mut inner = self.inner.borrow_mut();
+
+        if !inner.services.contains_key(&artifact) {
+            return Err(DeployError::FailedToDeploy);
+        }
+
+        if !inner.deployed.insert(artifact) {
+            return Err(DeployError::AlreadyDeployed);
+        }
+        Ok(())
+    }
+
+    fn check_deploy_status(&self, artifact: ArtifactSpec) -> Result<DeployStatus, DeployError> {
+        let artifact = if let ArtifactSpec::Rust(artifact) = artifact {
+            artifact
+        } else {
+            return Err(DeployError::WrongArtifact);
+        };
+
+        let inner = self.inner.borrow();
+
+        if inner.deployed.contains(&artifact) {
+            Ok(DeployStatus::Deployed)
+        } else {
+            Err(DeployError::FailedToDeploy)
+        }
+    }
+
+    fn init_service(
+        &mut self,
+        _: &mut EnvContext,
+        artifact: ArtifactSpec,
+        init: &InstanceInitData,
+    ) -> Result<(), InitError> {
+        let artifact = if let ArtifactSpec::Rust(artifact) = artifact {
+            artifact
+        } else {
+            return Err(InitError::WrongArtifact);
+        };
+
+        let mut inner = self.inner.borrow_mut();
+
+        if !inner.deployed.contains(&artifact) {
+            return Err(InitError::NotDeployed);
+        }
+
+        if inner.initialized.contains_key(&init.instance_id) {
+            return Err(InitError::ServiceIdExists);
+        }
+
+        let service = inner.services.remove(&artifact).unwrap();
+        inner.initialized.insert(init.instance_id, service);
+        Ok(())
+    }
+
+    fn execute(
+        &self,
+        context: &mut EnvContext,
+        dispatch: CallInfo,
+        payload: &[u8],
+    ) -> Result<(), ExecutionError> {
+        let inner = self.inner.borrow();
+        let instance = inner.initialized.get(&dispatch.instance_id).unwrap();
+
+        let ctx = TransactionContext::new(context, self);
```

Comment:
Why wouldn't include payload in `TransactionContext`? Does the payload relate to transaction here, doesn't it ?

Reply:
It is passed as a separate argument and converted to typed structure. Why would anyone need payload bytes?

Reply:
I mean to add `payload` as a field of `TransactionContext` struct. And then:
```rust
let ctx = TransactionContext::new(context, self, payload);
instance
      .call(dispatch.method_id, ctx)
      .expect("Dispatch error")
```

Logically a payload is more related to `TransactionContext` as for me. Because it's a part of the transaction, isn't it? But I could be wrong.

Reply:
Here is an example of usage of transaction context https://github.com/exonum/exonum/pull/1263/files#diff-b86eb31f9bf79bec1ba1ec769dc47638R37
`TransactionContext` structure passed as is, but payload is converted to `TimestampTx`. `TransactionContext` here is something other that transaction itself(payload). If we would add payload to Context we would have duplication of payload.

Reply:
Ok. I got it 👍🏻